src/test/tools/gran_detecto/gran_detecto.c

```diff
@@ -0,0 +1,356 @@
+
+#include <assert.h>
+#include <errno.h>
+#include <getopt.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include "libpmem2.h"
+#include "file.h"
+#include "os.h"
+#include "util.h"
+
+#define KILOBYTE (1 << 10)
+#define NOT_SET_FD (-1)
+
+enum gran_detecto_mode {
+	NOT_SET,
+	VALIDATE,
+	DETECT,
+};
+
+/*
+ * tool_ctx -- essential parameters used by gran_detecto
+ */
+struct tool_ctx
+{
+	enum pmem2_granularity actual_granularity;
+	char *probe_file_path;
+	int fd;
+	enum gran_detecto_mode state;
+
+	/* user arguments */
+	enum pmem2_granularity expected_granularity;
+	char *path;
+};
+
+/*
+ * print_usage -- print short description of usage
+ */
+static void
+print_usage(void)
+{
+	printf("Usage: gran_detecto -h\n"
+	"       gran_detecto <-b|-c|-d|-p> <path>\n"
+	"Available options:\n"
+	"-b, --byte          - check if <path> has byte granularity\n"
+	"-c, --cache-line    - check if <path> has cache line granularity\n"
+	"-d, --detect        - detect the smallest available granularity for <path>\n"
+	"-p, --page          - check if <path> has page granularity\n"
+	"-h, --help          - print this usage info\n");
+}
+
+/*
+ * long_options -- command line options
+ */
+static const struct option long_options[] = {
+	{"byte", no_argument, NULL, 'b'},
+	{"cache-line", no_argument, NULL, 'c'},
+	{"detect", no_argument, NULL, 'd'},
+	{"page", no_argument, NULL, 'p'},
+	{"help", no_argument, NULL, 'h'},
+	{NULL, 0, NULL, 0},
+};
+
+/*
+ * parse_args -- parse command line arguments
+ */
+static int
+parse_args(int argc, char *argv[], struct tool_ctx *ctx)
+{
+	/* tool context initialization */
+	ctx->expected_granularity = PMEM2_GRANULARITY_BYTE;
+	ctx->actual_granularity = PMEM2_GRANULARITY_BYTE;
+	ctx->fd = NOT_SET_FD;
+	ctx->state = NOT_SET;
+	ctx->probe_file_path = NULL;
+	ctx->path = NULL;
+
+	int opt;
+	while ((opt = getopt_long(argc, argv, "bcdhp", long_options, NULL)) !=
+		-1) {
+		switch (opt) {
+			case 'b':
+				ctx->state = VALIDATE;
+				ctx->expected_granularity =
+							PMEM2_GRANULARITY_BYTE;
+				break;
+			case 'c':
+				ctx->state = VALIDATE;
+				ctx->expected_granularity =
+						PMEM2_GRANULARITY_CACHE_LINE;
+				break;
+			case 'd':
+				ctx->state = DETECT;
+				break;
+			case 'p':
+				ctx->state = VALIDATE;
+				ctx->expected_granularity =
+							PMEM2_GRANULARITY_PAGE;
+				break;
+			case 'h':
+				print_usage();
+				exit(0);
+			default:
+				print_usage();
+				return 1;
+		}
+	}
+
+	if (ctx->state == NOT_SET) {
+		print_usage();
+		return 1;
+	}
+
+	if (optind < argc) {
+		ctx->path = argv[optind];
+	} else {
+		fprintf(stderr, "gran_detecto: path cannot be empty.\n");
+		print_usage();
+		return 1;
+	}
+
+	return 0;
+}
+
+/*
+ * cleanup_file -- close the file descriptor and remove the file and free file
+ * name variable if applicable
+ */
+#ifdef __linux__
+static void
+cleanup_file(struct tool_ctx *ctx)
```

Comment:
Again, this would be unnecessary if you simply used the same code for all platforms.